Softeq.XToolkit.Common/Commands/AsyncCommand.cs

```diff
@@ -13,65 +14,81 @@ namespace Softeq.XToolkit.Common.Commands
+        ///     at the moment due to the use of WeakActions (see http://stackoverflow.com/questions/25730530/).
         /// </param>
         /// <param name="canExecute">The Function that verifies whether or not AsyncCommand should execute.</param>
-        /// <param name="onException">If an exception is thrown in the Task, <c>onException</c> will execute.</param>
-        public AsyncCommand(Func<Task> execute, Func<bool> canExecute, Action<Exception>? onException = null)
-            : this(_ => execute(), _ => canExecute?.Invoke() ?? true, onException)
+        /// <param name="onException">If an exception is thrown in the Task, <c>onException</c> will be executed.</param>
+        public AsyncCommand(
+            Func<Task> execute,
+            Func<bool>? canExecute = null,
+            Action<Exception>? onException = null)
+            : base(onException)
         {
+            if (execute == null)
+            {
+                throw new ArgumentNullException(nameof(execute));
+            }
+
+            _execute = new WeakFunc<Task>(execute);
+
+            if (canExecute != null)
+            {
+                _canExecute = new WeakFunc<bool>(canExecute);
+            }
         }
 
         /// <summary>
-        ///     Initializes a new instance of the <see cref="AsyncCommand"/> class.
+        ///     Defines the method that determines whether the command can execute in its current state.
         /// </summary>
-        /// <param name="execute">
-        ///     The Function executed when Execute or ExecuteAsync is called.
-        ///     This does not check canExecute before executing and will execute even if canExecute is false.
+        /// <param name="parameter">
+        ///     Data used by the command. If the command does not require data
+        ///     to be passed, this object can be set to a null reference.
         /// </param>
-        /// <param name="canExecute">The Function that verifies whether or not AsyncCommand should execute.</param>
-        /// <param name="onException">If an exception is thrown in the Task, <c>onException</c> will execute.</param>
-        public AsyncCommand(
-            Func<object, Task> execute,
-            Func<object?, bool>? canExecute,
-            Action<Exception>? onException = null)
-            : base(canExecute)
+        /// <returns>true if this command can be executed; otherwise, false.</returns>
+        public bool CanExecute(object? parameter)
         {
-            _execute = execute;
-            _onException = onException;
+            if (!_execute.IsAlive)
+            {
+                return false;
+            }
+
+            if (IsRunning)
+            {
+                return false;
+            }
+
+            if (_canExecute == null)
+            {
+                return true;
+            }
+
+            if (!_canExecute.IsAlive)
+            {
+                return false;
+            }
+
+            return _canExecute.Execute();
```

Comment:
Yeap, the idea is that non-generic version of AsyncCommand does not actually use parameter in any way (similar to non-generic RelayCommand). If parameter is actually needed - it is possible to use AsyncCommand<object> or AsyncCommand<T> for a specific type

Reply:
I think in this case we should follow the `ICommand` interface to support `CanExecute` behavior:
```cs
bool CanExecute (object parameter)
```

Reply:
The idea was to support `Execute(object parameter)` also, as to provide `ICommand` interface.

In this case, we should support:
```cs
AsyncCommand(Func<object, Task> execute, Func<object, bool> canExecute)
```

Reply:
The thing is that
```
AsyncCommand(Func<object, Task> execute, Func<object, bool> canExecute)
```
is actually an exact duplication of 
```
AsyncCommand<T>(Func<T, Task> execute, Func<T, bool> canExecute)
```
with `object` as type parameter. So, in this case, it seems like non-generic AsyncCommand is actually not needed at all

Reply:
Yes, it's a good point.

but, we would like to avoid such a situation:

```cs
ICommand command = new AsyncCommand(() => ..., (?) => ... );

// ...

command.Execute(someParameter)    // <-- We need a notification about 
command.CanExecute(someParameter) //     wrong behavior in those cases
```

Reply:
You mean, if we try to use non-generic commands with parameters, different from null? I guess we could add another debug message in this case, like
`"Command do not use parameter, but was provided with not-null value of type {parameter.GetType()}"`

Reply:
Yes, it may sense in case when RelayCommand was used as ICommand, and then was replaced to AsyncCommand

Yes, agree with debug message

Reply:
OK, I will add a message. 
Just a small note thought: non-generic RelayCommand doesn't use parameters in any way too. So the message will appear only if non-generic implementations (RelayCommand/AsyncCommand) would be replaced with generic ones (RelayCommand<T>/AsyncCommand<T>)